Frameworks/Foundation/NSScanner.mm

```diff
@@ -575,118 +557,283 @@ - (BOOL)scanUpToString:(NSString*)string intoString:(NSString**)intoString {
+
+                } else if (unicode >= 'a' && unicode <= 'f') {
+                    if (scannedPeriod) {
+                        mantissa += (mantissaDigitPlace /= 16) * (unicode - 'a' + 10);
+                    } else {
+                        mantissa = (mantissa * 16) + (unicode - 'a' + 10);
+                    }
+                    hasValue = true;
+
+                } else if (unicode >= 'A' && unicode <= 'F') {
+                    if (scannedPeriod) {
+                        mantissa += (mantissaDigitPlace /= 16) * (unicode - 'A' + 10);
+                    } else {
+                        mantissa = (mantissa * 16) + (unicode - 'A' + 10);
+                    }
+                    hasValue = true;
+
+                } else if (unicode == 'p' || unicode == 'P') {
+                    state = STATE_P;
+
+                } else {
+                    goto done_scanning_hex_double;
+                }
+                break;
+
+            case STATE_P:
+                if (!hasValue) {
+                    goto done_scanning_hex_double;
+                }
+                if (unicode == '-') {
+                    state = STATE_EXPONENT;
+                    exponentIsNegative = true;
+
+                } else if (unicode == '+') {
+                    state = STATE_EXPONENT;
+
+                } else if (unicode >= '0' && unicode <= '9') {
+                    exponent = (unicode - '0');
+                    state = STATE_EXPONENT;
+
+                } else {
+                    goto done_scanning_hex_double;
+                }
+                break;
+
+            case STATE_EXPONENT:
+                if (unicode >= '0' && unicode <= '9') {
+                    // can stop calculating the change to the exponent once this is false, and let ldexp return 0 or HUGE_VAL
+                    if (exponent <= (exponentIsNegative ? 1023 : 1024)) {
+                        exponent = exponent * 10 + (unicode - '0');
+                    }
+
+                } else {
+                    goto done_scanning_hex_double;
+                }
+                break;
+        }
+    }
+
+done_scanning_hex_double:; // Allows breaking out of the loop from inside the switch
+    if (hasValue) {
+        if (result) {
+            if (exponentIsNegative) {
+                exponent *= -1;
+            }
+
+            *result = mantissaIsNegative ? -ldexp(mantissa, exponent) : ldexp(mantissa, exponent);
+        }
+
+        _scanLocation = i;
+        return YES;
+    }
+
+    return NO;
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (BOOL)scanHexFloat:(float*)result {
-    UNIMPLEMENTED();
-    return StubReturn();
+    double doubleValue;
+    if (![self scanHexDouble:&doubleValue]) {
+        return NO;
+    }
+
+    if (result) {
+        *result = static_cast<float>(doubleValue);
+    }
+    return YES;
 }
 
 /**
  @Status Interoperable
  @Notes
 */
 - (BOOL)scanUnsignedLongLong:(unsigned long long*)pValue {
```

Comment:
`unsigned` and `long` are just modifiers, not actual types, so saying `unsigned long long` makes the compiler infer the type which by default is `int`.  It's common to elide the int, but I'm not a fan 🐋